feat(AIP-121) Add checklist of resource-oriented APIs

toumorokoshi opened the pull request:
To help clarify resource-oriented requirements,
enumerating a list of requirements for
resource-oriented APIs.

Note that this includes on requirement from #993, and should be sequenced accordingly. If this is approved before #993 I can remove that line.

toumorokoshi committed:
Add checklist of resource-oriented APIs

To help clarify resource-oriented requirements,
enumerating a list of requirements for
resource-oriented APIs.

bgrant0607 commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+
+- The resource must have a unique [resource name][] by which it can be
+  identified and referenced.
+- The resource exposes only standard methods ([Get][], [List][], [Create][],
```
Where these bullets correspond to sections below, it may be useful to link to those sections.

On this specific bullet, is it intended to require or preclude anything? For example, does it preclude third-party data-plane protocols?

noahdietz replied:
I am also a little confused by the intention of this point.

toumorokoshi replied:
It was just trying to aggregate the underlying sections, but it is applying a restriction.

Custom methods as defined by AIP 136 to provide some restrictions around the formatting of the path, methods that should be used, etc. 

I guess conversely it's not *that* restrictive anyway, since custom methods are a bunch of shoulds today. 

If you don't think it's meaningful to call out the categories of methods you're limited to, I can remove the line.

toumorokoshi replied:
removed.

noahdietz commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+
+For a valid resource-oriented API, the following must be satisfied:
+
+- The resource must have a unique [resource name][] by which it can be
```
```suggestion
- Each resource must have a unique [resource name][] by which it can be
```

There can be more than one resource in a given API.

noahdietz replied:
```suggestion
- The resource must define a unique [resource name][] pattern or patterns by which it can be
```

Might want to check the line length of the resulting change.

noahdietz commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+
+- The resource must have a unique [resource name][] by which it can be
+  identified and referenced.
+- The resource exposes only standard methods ([Get][], [List][], [Create][],
```
```suggestion
- Each resource exposes only standard methods ([Get][], [List][], [Create][],
```

noahdietz commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+  identified and referenced.
+- The resource exposes only standard methods ([Get][], [List][], [Create][],
+  [Update][], [Delete][]) or [custom methods][].
+- The resource API must support [Get][] and [List][] operations.
```
Copied from the other PR:
This language conflicts with Singleton Resources that must not support List https://google.aip.dev/156 so I'm not sure we can enforce "must" for List. I think it has to be the following:
```suggestion
- The resource API must support a [Get][] operation and should support a [List][] operation.
```

toumorokoshi replied:
I'll add the singleton exception, as decided in the other PR. let me know if you prefer something different.

noahdietz commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+- The resource exposes only standard methods ([Get][], [List][], [Create][],
+  [Update][], [Delete][]) or [custom methods][].
+- The resource API must support [Get][] and [List][] operations.
+- Requests must be stateless: any individual request is independent of
```
I would not use the term `stateless`, that directly conflicts with language in https://google.aip.dev/136#stateless-methods.

toumorokoshi replied:
This was actually lifted from the "stateless protocol" section in this doc:

```
### Stateless protocol

As with most public APIs available today, resource-oriented APIs **must**
operate over a stateless protocol: The fundamental behavior of any individual
request is independent of other requests made by the caller.

In an API with a stateless protocol, the server has the responsibility for
persisting data, which may be shared between multiple clients, while clients
have sole responsibility and authority for maintaining the application state.
```

to be honest I feel like the guidance in this section is a little strange: a resource-oriented API necessitates some storage of the resource itself. I guess the point is that a single request doesn't depend on some other request, but IMO that's the default assumption vs something that needs to be called out specifically.

What do you think? Is it worth filing an issue to note some cleanup here?

noahdietz replied:
> but IMO that's the default assumption vs something that needs to be called out specifically.
>
> What do you think? Is it worth filing an issue to note some cleanup here?

Nah, I don't think it's worth a change atm. I'd rather we be explicit about the "stateless protocol" thing. If anything, I'd say perhaps we need clarification that this doesn't mean the same thing as AIP-136 Stateless Methods haha

> This was actually lifted from the "stateless protocol" section in this doc:

Ah indeed. Can you use "stateless protocol" consistently here then?

toumorokoshi replied:
SGTM! done.

noahdietz reviewed (commented):
Anywhere in this list that `must` or `should` are used must (no pun intended) be bold text

toumorokoshi committed:
address feedback

- fixing formatting (e.g. bolding of must)
- removing the CRUDL + custom methods restriction, as the
  terminology was not clear and very little is actually limited
  by adding that line.
- updating guidance for Get and List to match #993.
- fixing wording of stateless to clarify referring to stateless protocols.

toumorokoshi committed:
Merge branch 'master' into toum/resource-requirements

shwoodard commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+- Each resource **must** have a unique [resource name][] by which it can be
+  identified and referenced.
+- The resource API must support [Get][].
+- The resource API must support [List][], with the exception of [singleton
+  resources][].
```
Why must it support List? Sometimes the datasets are large, so large that pagination becomes a complex problem or other business reasons may make it so that List is not imperative. I agree that Get should be required, but why List?

toumorokoshi replied:
This is from #993, @bgrant0607 requested this integration.

I think this was long the lines of normalizing as many operations as possible. 

If a list wasn't supported though, it'd be fairly problematic. For GCP,  how would GCloud do list? or Console show all of the resource?

bgrant0607 replied:
+1. We should really be driving towards more requirements and away from flexibility.

Anyway, from what I've seen, List is more commonly supported than Get currently.

shwoodard commented on aip/general/0121.md:
```diff
@@ -53,6 +53,19 @@ database schema. In fact, having an API that is identical to the underlying
+- The resource API must support [Get][].
+- The resource API must support [List][], with the exception of [singleton
+  resources][].
+- Resources must operate over a stateless protocol: any individual request is
```
Can we just say "in a stateless manor"? A protocol is like `tcp` or `http` afaict and the word, doesn't quite fit here.
